### PR TITLE
Add nats streaming output plugin

### PR DESCRIPTION
**Pull request by sixlettervariables**
This adds [NATS Streaming](https://github.com/nats-io/go-nats-streaming) support as an output option (mirroring the existing NATS output plugin).

I broke out the godeps update into its own commit, as there were more changes than just the addition of go-nats-streaming, and I'm not experienced enough with go to judge the correctness of the other changes.

### Required for all PRs:

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

### CODE REVIEW

File: plugins/outputs/stan/stan.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	Subject string
+
+	// Path to CA file
+	SSLCA string `toml:"ssl_ca"`
+	// Path to host cert file
+	SSLCert string `toml:"ssl_cert"`
+	// Path to cert key file
+	SSLKey string `toml:"ssl_key"`
+	// Use SSL but skip chain & host verification
+	InsecureSkipVerify bool
```
Comment (bruth) [inline plugins/outputs/stan/stan.go]: Missing `toml:"insecure_skip_verify"`.

File: plugins/outputs/stan/stan.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
+package stan
+
+import (
+	"fmt"
+
+	nats_client "github.com/nats-io/go-nats"
+	stan_client "github.com/nats-io/go-nats-streaming"
```
Comment (bruth) [inline plugins/outputs/stan/stan.go]: I suggest just `nats` and `stan` as package names. This is more idiomatic in Go. If you want to retain the "client" part, then use `natsClient` and `stanClient`.

File: plugins/outputs/stan/stan.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+	// override authentication, if any was specified
+	if n.Username != "" {
+		natsOpts.User = n.Username
+		natsOpts.Password = n.Password
+	}
+
+	// override TLS, if it was specified
+	tlsConfig, err := internal.GetTLSConfig(
+		n.SSLCert, n.SSLKey, n.SSLCA, n.InsecureSkipVerify)
```
Comment (bruth) [inline plugins/outputs/stan/stan.go]: If you are going add a line break in a call like this, put one argument per line:

```go
tlsConfig, err := internal.GetTLSConfig(
    n.SSLCert,
    n.SSLKey,
    n.SSLCA,
    n.InsecureSkipVerify,
)
```

File: plugins/outputs/stan/stan.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	// try and connect to the NATS server
+	natsConn, err := natsOpts.Connect()
+	if err != nil {
+		return err
+	}
+
+	// try and connect to the NATS Streaming cluster
+	n.conn, err = stan_client.Connect(n.ClusterID, n.ClientID, stan_client.NatsConn(natsConn))
```
Comment (bruth) [inline plugins/outputs/stan/stan.go]: If you are going to break arguments by line above, I suggest doing the same here for consistency.

File: plugins/outputs/stan/stan.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+	return err
+}
+
+func (n *STAN) Close() error {
+	n.conn.Close()
```
Comment (bruth) [inline plugins/outputs/stan/stan.go]: This returns an error, so this could be returned.

File: plugins/outputs/stan/stan.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		if err != nil {
+			return err
+		}
+
+		err = n.conn.Publish(n.Subject, buf)
+		if err != nil {
+			return fmt.Errorf("FAILED to send NATS Streaming message: %s", err)
+		}
+	}
+	return nil
```
Comment (bruth) [inline plugins/outputs/stan/stan.go]: Add empty line above to break up the block.

File: plugins/outputs/stan/stan.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	nats_client "github.com/nats-io/go-nats"
+	stan_client "github.com/nats-io/go-nats-streaming"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+	"github.com/influxdata/telegraf/plugins/serializers"
+)
+
+type STAN struct {
```
Comment (bruth) [inline plugins/outputs/stan/stan.go]: This is more preference, but could just be called `Stan` rather than all caps.

**Comment by danielnelson** [comment]
Couple high level questions: what about naming the plugin `nats`? Usually we would have a `nats` output and a `nats_consumer` input for send/recv, plus a `nats` input that gathers metrics about the service itself.

Secondly, why add the new dependency, since can't we publish with go-nats?

File: plugins/outputs/stan/stan.go (hunk 3 above)
Comment (sixlettervariables) [reply plugins/outputs/stan/stan.go]: Is there a recommended line length limit for the repo?

**Comment by sixlettervariables** [comment]
@bruth: I appreciate the detailed comments. New to go, trying to pick up the idioms.

@danielnelson: I'm not opposed to that. NATS and NATS Streaming are two separate protocols, so I split it out in my PR. (I also don't represent the NATS team, who may have another opinion)

**Comment by bruth** [comment]
@danielnelson As @sixlettervariables said, they are two different protocols, but NATS Streaming is built on top of NATS which is why the two dependencies exist. NATS is a message broker with an *at-most-once* guarantee. If consumers are offline, messages are dropped. NATS Streaming is an *at-least-once* guarantee and messages are durable (like Kafka) so consumers can consume and resume when they are available.

**Comment by sixlettervariables** [comment]
I reviewed the prior ci failures and they appear to revolve around the updates to godeps. To produce the file I ran:
```sh
$ gdm restore
$ GOOS=linux gdm save
```
Should I revert all of the deltas but the addition of go-nats-streaming to the file?

**Comment by danielnelson** [comment]
> Should I revert all of the deltas but the addition of go-nats-streaming to the file?

Yes please, sorry I'll sort out this file soon.

**Comment by lcharkiewicz** [comment]
@danielnelson any plans for merging this PR soon (like till 1.8 release)?

**Comment by danielnelson** [comment]
@lcharkiewicz I don't have a firm spot in the schedule yet.

**Comment by srebhan** [comment]
It has been a while now since last activity. If someone wants to pick this up, please reopen this PR!

**Comment by sixlettervariables** [comment]
@srebhan we gave up on getting it merged and went with a different product. It was ready to go as far as I can recall.

**Pull request closed without merging**